packages/block-editor/src/components/use-on-block-drop/index.js

```diff
@@ -212,27 +212,15 @@ export function onHTMLDrop(
  * @return {Object} An object that contains the event handlers `onDrop`, `onFilesDrop` and `onHTMLDrop`.
  */
 export default function useOnBlockDrop( targetRootClientId, targetBlockIndex ) {
+	const hasUploadPermissions = useSelect(
+		( select ) => select( blockEditorStore ).getSettings().mediaUpload,
+		[]
+	);
 	const {
 		canInsertBlockType,
 		getBlockIndex,
 		getClientIdsOfDescendants,
-		hasUploadPermissions,
-	} = useSelect( ( select ) => {
-		const {
-			canInsertBlockType: _canInsertBlockType,
-			getBlockIndex: _getBlockIndex,
-			getClientIdsOfDescendants: _getClientIdsOfDescendants,
-			getSettings,
-		} = select( blockEditorStore );
-
-		return {
-			canInsertBlockType: _canInsertBlockType,
```

Comment:
To have the full picture, this could also be refactored as:

```js
return {
    ...pick( [ 'canInsertBlockType, getBlockIndex, getClientIdsOfDescendants ),
    hasUploadPermissions: getSettings().mediaUpload,
};
```

Reply:
It's doesn't solve my main problem: it's awkward to return selector functions in a mapping.

Reply:
Right, it's a workaround. I think for `withDispatch` we were providing an 3rd argument so it could use selectors without passing it down from `withSelect`. Here it won't work so your exploration is really worth conisdering.